src/Castle.Core.Tests/ExcludeOnClrAttribute.cs

```diff
@@ -0,0 +1,44 @@
+﻿// Copyright 2004-2018 Castle Project - http://www.castleproject.org/
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.f
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle
+{
+	using System;
+
+	using NUnit.Framework;
+	using NUnit.Framework.Interfaces;
+	using NUnit.Framework.Internal;
+
+	[AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = false)]
+	public sealed class ExcludeOnClrAttribute : NUnitAttribute, IApplyToTest
+	{
+		private static bool IsRunningOnMono() => Type.GetType("Mono.Runtime", throwOnError: false) != null;
+
+		private string reason;
+
+		public ExcludeOnClrAttribute(string reason)
+		{
+			this.reason = reason;
+		}
+
+		public void ApplyToTest(Test test)
+		{
+			if (test.RunState != RunState.NotRunnable && test.RunState != RunState.Skipped && !IsRunningOnMono())
```

Comment:
This is actually a bit hacky. Just because we're *not* running on Mono doesn't mean that we're running on the CLR. This is good enough for the current build setup, but not ideal.

Reply:
Oh yer, I saw that and thought this needs to be better, will give it some thought tomorrow.

Reply:
I was looking at dotnet/buildtools' `[SkipOnTargetFramework]` for inspiration, and asked myself whether it would be an option for Castle.Core to drop all magic OS & runtime guessing and have the necessary parameters passed in from the command line / build script. (There would have to be some reasonable default, which might be tricky. Some people might forget to pass the parameters. Or someone might execute the tests using a different runner.)

Reply:
Just discovered [`RuntimeInformation.FrameworkDescription`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.frameworkdescription), which appears to work on .NET, .NET Core, and Mono. How about reimplementing a single `[ExcludeOnRuntime(<runtime id>)]` based on that? In a first iteration, it would perhaps be version-agnostic, that could still be added later on.

Apart from that, BenchmarkDotNet might be another good place to find runtime / OS detection logic.

And finally, there's always the option of dropping all runtime / OS detection magic and having the required parameters passed in from the build script (but that seems tricky for a couple reasons).

Reply:
I agree that it would be preferred to do at runtime rather than pass a parameter into the runner so we don't run into problems with using different test runners.

BenchmarkDotNet treats "CLR" as just the full/desktop CLR, and not also the CoreCLR. A single attribute looks like a good option as we might want to enable some on the CoreCLR once things get fixed.

Regarding the attribute name though, should it be "runtime" or "framework"? For example, the runtime is CoreCLR but heaps of the code we depend on is in CoreFX, so in general the framework is .NET Core? If we go with "runtime" we could probably do something like this, maybe even drop "runtime" from the attribute name as you'd always have the enum:

```
[ExcludeOnRuntime(Runtime.Mono)]
[ExcludeOnRuntime(Runtime.Clr)]
[ExcludeOnRuntime(Runtime.CoreClr)]
[ExcludeOnRuntime(Runtime.Clr | Runtime.CoreClr)]
[ExcludeOnRuntime(Runtime.ClrAndCoreClr)] // I think I prefer the bitwise one
```

With respect to determining which runtime/framework the tests are running on I don't mind requiring .NET 4.7.1 (supports Windows 7 SP1) for the unit tests, that is what BenchmarkDotNet is using for their shipping code anyway, and falling back to conditional compilation.

Reply:
> Regarding the attribute name though, should it be "runtime" or "framework"?

Of the two, I'm starting to tend more towards "framework", but I'm still undecided. I'll happily defer to your wise decision.

Perhaps, it doesn't matter too much, since any particular framework usually comes with its own runtime, so the two are tied together. I don't know of any two CLI runtimes that share a common framework library. Framework is the more general term of the two, so probably less restrictive if we wanted to enhance that `[Exclude...]` attribute later on.

dotnet/corefx for one prefers the term "framework": see e.g. its [`[SkipOnTargetFramework]` attribute](https://github.com/dotnet/corefx/blob/ccdb5a48a49d3e932aa00d02749777464f4c8d5b/Documentation/project-docs/developer-guide.md#skipontargetframeworkattribute) and the [`TargetFrameworkMoniker` flags enum](https://github.com/dotnet/buildtools/blob/4918a5245e955480db77a071390b91698e0c820b/src/xunit.netcore.extensions/TargetFrameworkMonikers.cs). Not sure I like it the mix-up of runtime and version into a single enumeration, though.

"Platform" would be another possibility. It's somewhat vague but that could actually be good for an attribute that we might want to expand with additional constraints later on.

> ```csharp
> [ExcludeOnRuntime(Runtime.ClrAndCoreClr)] // I think I prefer the bitwise one
> ```

Agreed, a flags enum would be nicer.

Reply:
Let's go with framework, the framework is the distribution of a runtime and BCL. For example, UWP's runtime is CoreCLR but is a different framework.

Which option do you like for the attribute name? `[Framework]` would work more like NUnit's `[Platform]`?

```csharp
[ExcludeOnFramework(Framework.Mono)]
[ExcludeOn(Framework.Mono)]
[Framework(Exclude = Framework.Mono)]

[Framework(Framework.Mono)] // could be Mono only

[Framework(~Framework.Mono)] // this should be doable, but is it clear?
```

Reply:
I'd go with the first one iff the `Framework` parameter is positional and not an optional, named one. Otherwise, I'd go with your second suggestion, which would allow for more future extensibility (e.g. `[ExcludeOn(OS = OS.Unix, Framework = Framework.NetCore, ...)]`

(The third suggestion suffers from `Exclude` being a named parameter and therefore IIRC optional. What effect would `[Framework()]` have? The last two suggestions suffer from unclear intent: Does a `[Framework]` attribute imply inclusion or exclusion?)

Reply:
Good points. I'm also heavily leaning on the first one now too, very explicit and we don't really have (or want) unit tests that can't be run everywhere, obviously except for framework defects.

Would you now support multiple declarations on the same member (like the example below), or just merge text into one?

```csharp
[ExcludeOnFramework(Framework.NetCore, "Blah blah see dotnet/coreclr/issues/1")]
[ExcludeOnFramework(Framework.NetFramework, "Other see msconnect/2")]
public void MyTest() { ... }
```